### PR TITLE
Sssp win fix

### PR DESCRIPTION
**Pull request by orrrrtem**
Reimplemented optional results id init.
Fixed warnings.
Removed parts of duplicated code.

### CODE REVIEW

File: cpp/oneapi/dal/algo/shortest_paths/detail/traverse_default_kernel.hpp [hunk 1]
```diff
@@ -99,7 +99,13 @@ inline void relax_edges(const Topology& t,
         const EdgeValue new_dist = dist[u].load() + v_w;
         while (new_dist < old_dist) {
             if (dist[v].compare_exchange_strong(old_dist, new_dist)) {
-                std::int64_t dest_bin = new_dist / delta;
+                ONEDAL_ASSERT(new_dist > 0);
+                ONEDAL_ASSERT(delta > 0);
+                ONEDAL_ASSERT(new_dist / delta <= static_cast<EdgeValue>(std::min(
+                                                      std::numeric_limits<std::int64_t>::max(),
+                                                      std::numeric_limits<EdgeValue>::max())));
+                const std::int64_t dest_bin = static_cast<std::int64_t>(new_dist / delta);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/shortest_paths/detail/traverse_default_kernel.hpp:107]: Looks like code duplication

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel_nrh.cpp [hunk 2]
```diff
@@ -80,7 +86,13 @@ inline void relax_edges_with_pred(const Topology& t,
         const EdgeValue new_dist = dp[u].dist + v_w;
         if (new_dist < old_dp.dist) {
             dp[v] = dist_pred(new_dist, u);
-            std::int64_t dest_bin = new_dist / delta;
+            ONEDAL_ASSERT(new_dist > 0);
+            ONEDAL_ASSERT(delta > 0);
+            ONEDAL_ASSERT(new_dist / delta <=
+                          static_cast<EdgeValue>(std::min(std::numeric_limits<std::int64_t>::max(),
+                                                          std::numeric_limits<EdgeValue>::max())));
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel_nrh.cpp:93]: Why don't you use separate `constexpr` value to store max distance?

File: cpp/oneapi/dal/algo/shortest_paths/detail/traverse_default_kernel.hpp (hunk 1 above)
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/detail/traverse_default_kernel.hpp:107]: fixed

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel_nrh.cpp (hunk 2 above)
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel_nrh.cpp:93]: It is used only in asserts that is in debug mode. Complicating code with ifdef assert does not cost such perf improvements for debugging.

**Review by bysheva — APPROVED** [review]

**Pull request merged**